tinyiiod.c

```diff
@@ -62,160 +74,219 @@ int tinyiiod_read_command(struct tinyiiod *iiod)
 void tinyiiod_write(struct tinyiiod *iiod, const char *data, size_t len)
 {
-	iiod->ops->write(data, len);
+	iiod->ops->write(iiod->instance_id, data, len);
 }
 
+/***************************************************************************//**
+ * @brief tinyiiod_write_string
+*******************************************************************************/
 void tinyiiod_write_string(struct tinyiiod *iiod, const char *str)
 {
 	tinyiiod_write(iiod, str, strlen(str));
 }
 
-void tinyiiod_write_value(struct tinyiiod *iiod, int value)
+/***************************************************************************//**
+ * @brief tinyiiod_write_value
+*******************************************************************************/
+void tinyiiod_write_value(struct tinyiiod *iiod, int32_t value)
 {
 	char buf[16];
 
-	snprintf(buf, sizeof(buf), "%i\n", value);
+	snprintf(buf, sizeof(buf), "%"PRIi32"\n", value);
 	tinyiiod_write_string(iiod, buf);
 }
 
+/***************************************************************************//**
+ * @brief tinyiiod_write_xml
+*******************************************************************************/
 void tinyiiod_write_xml(struct tinyiiod *iiod)
 {
 	size_t len = strlen(iiod->xml);
-
 	tinyiiod_write_value(iiod, len);
 	tinyiiod_write(iiod, iiod->xml, len);
 	tinyiiod_write_char(iiod, '\n');
 }
 
+/***************************************************************************//**
+ * @brief tinyiiod_do_read_attr
+*******************************************************************************/
 void tinyiiod_do_read_attr(struct tinyiiod *iiod, const char *device,
-		const char *channel, bool ch_out, const char *attr, bool debug)
+			   const char *channel, bool ch_out, const char *attr, bool debug)
 {
-	char buf[128];
-	ssize_t ret;
-
-	if (channel)
-		ret = iiod->ops->ch_read_attr(device, channel,
-				ch_out, attr, buf, sizeof(buf));
-	else
-		ret = iiod->ops->read_attr(device, attr,
-				buf, sizeof(buf), debug);
-
+	char buf[0x1000];
+	ssize_t ret = -ENODEV;
+	if(debug == 0) {
+		if (channel)
+			ret = iiod->ops->ch_read_attr(device, channel,
+						      ch_out, attr, buf, sizeof(buf));
+		else
+			ret = iiod->ops->read_attr(device, attr,
+						   buf, sizeof(buf), debug);
+	}
 	tinyiiod_write_value(iiod, (int) ret);
 	if (ret > 0) {
-		tinyiiod_write(iiod, buf, (size_t) ret);
-		tinyiiod_write_char(iiod, '\n');
+		buf[ret] = '\n';
+		tinyiiod_write(iiod, buf, (size_t) ret + 1);
 	}
 }
 
+/***************************************************************************//**
+ * @brief tinyiiod_do_write_attr
+*******************************************************************************/
 void tinyiiod_do_write_attr(struct tinyiiod *iiod, const char *device,
-		const char *channel, bool ch_out, const char *attr,
-		size_t bytes, bool debug)
+			    const char *channel, bool ch_out, const char *attr,
+			    size_t bytes, bool debug)
 {
-	char buf[128];
+	char buf[2048] = {0};
```

Comment:
To eliminate this large buffer we need a mechanism to separate the message beforehand and just give chunks of messages where we limit the size (ex: split the message in 128 bytes chunks and send multiple times). This is just an idea not sure of all the implications.